ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,233 @@ type clientHandshakeState struct {
 	session         *ClientSessionState
 }
 
+type CacheKeyGenerator interface {
+	Key(net.Addr) string
+}
+
+type ClientFingerprintConfiguration struct {
+	// Version in the handshake header
+	HandshakeVersion uint16
+
+	// if len == 32, it will specify the client random
+	// Otherwise, the field will be random
+	ClientRandom []byte
+
+	// if RandomSessionID > 0, will overwrite SessionID w/ that many
+	// random bytes when a session resumption occurs
+	RandomSessionID int
+	SessionID       []byte
+
+	// These fields will appear exactly in order in the ClientHello
+	CipherSuites       []uint16
+	CompressionMethods []uint8
+	Extensions         []ClientExtension
+
+	// Optional, both must be non-nil, or neither.
+	// Custom Session cache implementations allowed
+	SessionCache ClientSessionCache
+	CacheKey     CacheKeyGenerator
+}
+
+type ClientExtension interface {
+	// Produce the bytes on the wire for this extension, type and length included
+	Marshal() []byte
+
+	// Function will return an error if zTLS does not implement the necessary features for this extension
+	CheckImplemented() error
```

Comment:
In the Hash and Signature Algorithm Extension, if a algorithm combination listed to be offered by the client is not actually in zTLS. Or even a non-defined value.